### ISSUE TITLE
How to get a newer maven version alongside JDK 17

### ISSUE THREAD

**Issue by ghoshbishakh** [actions/setup-java#685]
**Description:**

After setting up JDK 17, the maven version we get is `3.8.8`
We want to get the latest maven version, `3.9.x`

Here is the workflow:

```yml
jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Set up JDK
        uses: actions/setup-java@v4
        with:
          java-version: '17'
          distribution: 'temurin'
      - run: mvn --version

```

**Task version:**
v4

**Platform:**
- [x] Ubuntu
- [ ] macOS
- [ ] Windows

**Runner type:**
- [x] Hosted
- [ ] Self-hosted

**Comment by aparnajyothi-y** [comment]
Hello @ghoshbishakh, Thank you for creating this issue and we will look into it :)

**Comment by gowridurgad** [comment]
Hi @ghoshbishakh, To use Maven version 3.9.9 alongside JDK 17, Maven needs to be manually installed since the ubuntu-latest runner image comes with Maven 3.8.8 by default.  For reference, please visit  https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2204-Readme.md. 
Here is an updated workflow configuration:

```
jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Install latest Maven
        run: |
          MAVEN_VERSION=3.9.9
          wget https://downloads.apache.org/maven/maven-3/$MAVEN_VERSION/binaries/apache-maven-$MAVEN_VERSION-bin.tar.gz
          tar xzvf apache-maven-$MAVEN_VERSION-bin.tar.gz
          sudo mv apache-maven-$MAVEN_VERSION /opt/maven
          sudo rm -f /usr/bin/mvn  # Remove existing symbolic link if it exists
          sudo ln -s /opt/maven/bin/mvn /usr/bin/mvn  # Create new symbolic link
      - name: Set up JDK
        uses: actions/setup-java@v4
        with:
          java-version: '17'
          distribution: 'temurin' 
      - run: mvn --version
```

**Comment by ghoshbishakh** [comment]
Thanks a lot @gowridurgad . I will try it out.

**Comment by jglick** [comment]
I think this should be reopened: 3.8.8 is not compatible with Java 17-based tools (fixed in https://github.com/apache/maven/pull/1286 in 3.9.6), so it is important to be running a newer version.

**Comment by gowridurgad** [comment]
Hi @jglick, We are able to use Maven version 3.8.8 alongside JDK 17.  Attached a screenshot for your reference. However, if you need to use Maven 3.9.x, you can manually install it in your GitHub Actions workflow. 

<img width="1422" alt="Screenshot 2024-09-27 at 4 21 48 PM" src="https://github.com/user-attachments/assets/d3ce2933-5f31-4ec6-b7f4-15d7cf7ab83e">

**Comment by timja** [comment]
> Hi @jglick, We are able to use Maven version 3.8.8 alongside JDK 17. Attached a screenshot for your reference. However, if you need to use Maven 3.9.x, you can manually install it in your GitHub Actions workflow.

The difference here will be a Maven plugin compiled with Java 17. Before the fix in 3.9.6 Maven could build Java 17 projects but couldn't use Maven plugins with bytecode newer than Java 14.

**Comment by kwin** [comment]
Using [Maven Wrapper](https://maven.apache.org/wrapper/) is the easiest way to enforce a specific Maven version on most operating systems.
